### PR TITLE
Secure and restrict REST-API connection

### LINKED ISSUE #41: Authenticated secure connection to the controller
Linstor Controller allows [restrict client access using client certificate](https://linbit.com/drbd-user-guide/linstor-guide-1_0-en/#s-rest-api-https-restricted-client).

Currently this will not work with linstor-proxmox plugin, so we need to have Linstor Controller REST API accessible from the whole network. This plugin doesn't even work with HTTPS connections (without restricting access by cert).

This means that anyone who has access to this network can use REST API to do whatever they want within the Linstor cluster, including gaining access to the virtual disks.

It would be great to be able to configure client certificate for each Proxmox node, so we can make our cluster more secure.

### PR DESCRIPTION
**Pull request by Stibila**
In default configuration LINSTOR REST-API/controller is accessible from
anywhere in the network without restriction. It can be hovever secured by
HTTPS with restricting client access by trusted client SSL certificate.

For this purpose new options were introduced:

apicrt - path to the trusted client certificate
apikey - path to the trusted client private key
apica - path to the CA certificate that signed REST-API HTTPS server cert

If both apicrt and apikey are configured in /etc/pve/storage.cfg the plugin
will change protocol to HTTPS and port to 3371 and provided certificates
and key will be used to establish connection to the REST-API.

Each node can use different key cert pair, but they must be stored in the
same location on each node.

Closes #41

### CODE REVIEW

**Comment by rck** [comment]
Hi @Stibila 

already thanks for the PR, I'm currently on vacation till 5th of July. If I did not comment till the end of that week feel free to kick me again.

Thanks so far, rck

**Comment by Stibila** [comment]
No problem. Enjoy vacation :)

File: LINSTORPlugin.pm [hunk 1]
```diff
@@ -18,6 +18,7 @@ use PVE::Storage;
 use PVE::Storage::Plugin;
 use PVE::JSONSchema qw(get_standard_option);
 
+
```
Comment (rck) [inline LINSTORPlugin.pm:21]: please remove, guess that extra line is some leftover.

File: LINSTORPlugin.pm [hunk 2]
```diff
@@ -103,6 +123,9 @@ sub options {
         content       => { optional => 1 },
         disable       => { optional => 1 },
         nodes         => { optional => 1 },
+        apicrt        => { optional => 1 },
+        apikey        => { optional => 1 },
+        apica        => { optional => 1 },
```
Comment (rck) [inline LINSTORPlugin.pm:128]: you might want to add an extra space to keep the alignment with the other options.

File: LINSTORPlugin.pm [hunk 3]
```diff
@@ -151,14 +174,53 @@ sub get_preferred_local_node {
...
 sub linstor {
     my ($scfg) = @_;
 
     my @controllers = split( /,/, get_controllers($scfg) );
 
+    my $apicrt = get_apicrt($scfg);
+    my $apikey = get_apikey($scfg);
+    my $apica = get_apica($scfg);
+
+    my %config = ();
```
Comment (rck) [inline LINSTORPlugin.pm:204]: unused, please remove

File: LINSTORPlugin.pm [hunk 4]
```diff
@@ -151,14 +174,53 @@ sub get_preferred_local_node {
...
+
+    my $proto = "http";
+    my $port = "3370";
+
+    # If cert an key are configured, change protocol and port
+    if ( defined $apicrt and $apikey ) {
```
Comment (rck) [inline LINSTORPlugin.pm:210]: `if ( defined $apicrt and defined $apikey )`?

**Review by rck — CHANGES_REQUESTED** [review]
All in all this LGTM. There are just some minor issues/leftovers. Please address these and force push. If you need help for that feel free to ask.

File: LINSTORPlugin.pm (hunk 2 above)
Comment (Stibila) [reply LINSTORPlugin.pm]: I somehow missed that. Fixed in commit e1c2916.

File: LINSTORPlugin.pm (hunk 3 above)
Comment (Stibila) [reply LINSTORPlugin.pm]: Fixed in commit e1c2916.

File: LINSTORPlugin.pm (hunk 4 above)
Comment (Stibila) [reply LINSTORPlugin.pm]: Totally newbie mistake. I'm embarrassed :) Fixed in commit e1c2916.

**Comment by Stibila** [comment]
Issues were fixed. Please review. Thanks.

**Comment by rck** [comment]
From what I see this is fine now. I will give it a try in my test cluster tomorrow. Feel free to squash the 2 commits and force push, otherwise I will fold them when I merge it.

**Comment by rck** [comment]
very well, then there is only one question left: how do I test it? :)
did you start from https://linbit.com/drbd-user-guide/linstor-guide-1_0-en/#s-linstor-rest-api-https ? And how did you then create the cert/key/ca? some step by step would be nice, including your `/etc/linstor/linstor.toml`

**Comment by Stibila** [comment]
I followed [Linstor guide](https://linbit.com/drbd-user-guide/linstor-guide-1_0-en/#s-linstor-rest-api-https) more or less.

Using keytool to create server and client certificates and then export server cert without key (configured on client as CA). Then converted client certs into PEM format, because I couldn't get REST::Client to work with pkcs12 for some reason. You can use same client certificate for each PVE node, but I decided to generate one for each node.

Here are my notes in the form of Bash commands:

```
# Execute on linstor controller:

linstor_keystore=/etc/linstor/ssl/keystore.jks
clients_keystore=/etc/linstor/ssl/clients.jks
linstorkeystore_pass=Test123.
clientkeystore_pass=Test12345.

# Linstor config
cat > /etc/linstor/linstor.toml < # For more options check documentation

[db]
  user = "linstor"
  password = "linstor"
  connection_url = "jdbc:h2:/var/lib/linstor/linstordb"

[http]
  enabled = true
  listen_addr = "::"
  port = 3370

[https]
  enabled = true
  listen_addr = "::"
  port = 3371

  # keystore containing the https server certificate
  keystore = "${linstor_keystore}"

  # keystore password to unlock the server certificate
  keystore_password = "${keystore_pass}"

  # to only allow clients with the correct certificates
  truststore = "${clients_keystore}"
  truststore_password = "${clientcert_pass}"
EOF

# Server certificate
mkdir /etc/linstor/ssl && chmod 750 /etc/linstor/ssl
keytool -keyalg rsa -keysize 4096 -genkey -keystore "${linstor_keystore}" -storepass "${keystore_pass}" -alias testlinstor.test.local -dname "CN=testlinstor.test.local, O=Test, L=Bratislava, C=SK"
chmod o-r "${linstor_keystore}"

# Extract CA
keytool -exportcert -storepass "${keystore_pass}" -alias testlinstor.test.local -keystore "${linstor_keystore}" -rfc -file /etc/linstor/ssl/ca.crt

# Client certificate for each node
keytool -keyalg rsa -keysize 4096 -genkey -keystore "${clients_keystore}" -storepass "${clientcert_pass}" -alias testvirthost1.test.local -dname "CN=testvirthost1.test.local, O=Test, L=Bratislava, C=SK"
keytool -keyalg rsa -keysize 4096 -genkey -keystore "${clients_keystore}" -storepass "${clientcert_pass}" -alias testvirthost2.test.local -dname "CN=testvirthost2.test.local, O=Test, L=Bratislava, C=SK"

# Export client certs to pkcs12
keytool -importkeystore -srckeystore "${clients_keystore}" -destkeystore /tmp/testvirthost1.p12 -srcstorepass ${clientcert_pass} -storepass ${clientcert_pass} -keypass ${clientcert_pass} -srcalias testvirthost1.test.local -srcstoretype jks -deststoretype pkcs12
keytool -importkeystore -srckeystore "${clients_keystore}" -destkeystore /tmp/testvirthost2.p12 -srcstorepass ${clientcert_pass} -storepass ${clientcert_pass} -keypass ${clientcert_pass} -srcalias testvirthost2.test.local -srcstoretype jks -deststoretype pkcs12

# Copy certs and CA to virthosts
scp /tmp/testvirthost1.p12 testvirthost1:/etc/linstor/ssl/client.p12
scp /etc/linstor/ssl/ca.crt testvirthost1:/etc/linstor/ssl/ca.crt
scp /tmp/testvirthost2.p12 testvirthost2:/etc/linstor/ssl/client.p12
scp /etc/linstor/ssl/ca.crt testvirthost2:/etc/linstor/ssl/ca.crt
rm /tmp/testvirthost*.p12

# Execute on both PVE nodes:
# Convert cert and key to PEM format without password
mkdir -p /etc/linstor/ssl/
openssl pkcs12 -in /etc/linstor/ssl/client.p12 -nokeys -out /etc/linstor/ssl/client.crt -passin "pass:${clientkeystore_pass}"
openssl pkcs12 -in /etc/linstor/ssl/client.p12 -nodes -nocerts -out /etc/linstor/ssl/client.key -passin "pass:${clientkeystore_pass}"

# Cleanup:
rm /etc/linstor/ssl/client.p12
chown root: -R /etc/linstor/ssl/
chmod o-r /etc/linstor/ssl/client.key
```

I hope this helps.

**Review by rck — APPROVED** [review]

**Pull request merged**